File: src/java/magmac/app/lang/node/FunctionSegments.java
package magmac.app.lang.node;

import magmac.api.collect.list.Lists;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.ExactRule;
import magmac.app.compile.rule.OrRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.StripRule;
import magmac.app.compile.rule.TypeRule;
import magmac.app.lang.Deserializers;
import magmac.app.lang.java.JavaDeserializers;
import magmac.app.lang.JavaRules;
import magmac.app.lang.LazyRule;
import magmac.app.lang.java.JavaFunctionSegment;

public final class FunctionSegments {
    public static CompileResult<JavaFunctionSegment> deserialize(Node node) {
        return Deserializers.orError("function-segment", node, Lists.of(
                Deserializers.wrap(JavaDeserializers::deserializeWhitespace),
                Deserializers.wrap(JavaDeserializers::deserializeComment),
                Deserializers.wrap(JavaDeserializers::deserializeFunctionStatement),
                Deserializers.wrap(JavaDeserializers::deserializeBlock),
                Deserializers.wrap(JavaDeserializers::deserializeReturn),
                Deserializers.wrap(JavaDeserializers::deserializeCase)
        ));
    }

    public static Rule initFunctionSegmentRule(LazyRule functionSegmentRule, Rule value, Rule definition) {
        var functionSegmentValueRule = FunctionSegmentValues.createFunctionSegmentValueRule(value, definition);

        Rule rule = new OrRule(Lists.of(
                new TypeRule("whitespace", new StripRule(new ExactRule(";"))),
                JavaRules.createTypedWhitespaceRule(),
                JavaRules.createCommentRule(),
                JavaRules.createStatementRule(functionSegmentValueRule),
                JavaRules.createBlockRule(functionSegmentRule, value, definition),
                JavaRules.createReturnRule(value),
                JavaRules.createCaseRule(value, functionSegmentRule)
        ));

        return functionSegmentRule.set(new StripRule("before", rule, ""));
    }

}

File: src/web/magmac/app/lang/node/FunctionSegments.ts
import { Lists } from "../../../../magmac/api/collect/list/Lists";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { ExactRule } from "../../../../magmac/app/compile/rule/ExactRule";
import { OrRule } from "../../../../magmac/app/compile/rule/OrRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { StripRule } from "../../../../magmac/app/compile/rule/StripRule";
import { TypeRule } from "../../../../magmac/app/compile/rule/TypeRule";
import { Deserializers } from "../../../../magmac/app/lang/Deserializers";
import { JavaDeserializers } from "../../../../magmac/app/lang/java/JavaDeserializers";
import { JavaRules } from "../../../../magmac/app/lang/JavaRules";
import { LazyRule } from "../../../../magmac/app/lang/LazyRule";
import { JavaCaseNode } from "../../../../magmac/app/lang/java/JavaCaseNode";
import { JavaFunctionSegment } from "../../../../magmac/app/lang/java/JavaFunctionSegment";
export class FunctionSegments {
	deserialize(node : Node) : CompileResult<JavaFunctionSegment> {return 0.orError( 0, 0, 0.of( 0.wrap( 0.deserializeWhitespace), 0.wrap( 0.deserializeFunctionStatement), 0.wrap( 0.deserializeBlock), 0.wrap( 0.deserializeReturn), 0.wrap( 0.deserialize)));;}
	initFunctionSegmentRule(functionSegmentRule : LazyRule, value : Rule, definition : Rule) : Rule {break;break;return 0.set( new StripRule( 0, 0, 0));;}
}
